sklearn/linear_model/ridge.py

```diff
@@ -1015,48 +1246,152 @@ def _errors_and_values_helper(self, alpha, y, v, Q, QT_y):
+        # X already centered
+        X_mean = np.zeros(X.shape[1], dtype=X.dtype)
+        if self.fit_intercept:
+            intercept = sqrt_sw[:, None]
+            X = np.hstack((X, intercept))
         # to emulate fit_intercept=True situation, add a column on ones
         # Note that by centering, the other columns are orthogonal to that one
         U, s, _ = linalg.svd(X, full_matrices=0)
         v = s ** 2
         UT_y = np.dot(U.T, y)
-        return v, U, UT_y
+        return X_mean, v, U, UT_y
+
+    def _errors_and_values_covariance_dense(
+            self, alpha, y, sqrt_sw, v, U, UT_y):
+        """Compute dual coefficients and diagonal of (Identity - Hat_matrix)
 
-    def _errors_and_values_svd_helper(self, alpha, y, v, U, UT_y):
-        """Helper function to avoid code duplication between self._errors_svd
-        and self._values_svd.
+        Used when we have an SVD decomposition of X
+        (n_features >= n_samples and X is dense).
         """
-        constant_column = np.var(U, 0) < 1.e-12
-        # detect columns colinear to ones
         w = ((v + alpha) ** -1) - (alpha ** -1)
-        w[constant_column] = - (alpha ** -1)
-        # cancel the regularization for the intercept
+        if self.fit_intercept:
+            # detect intercept column
+            normalized_sw = sqrt_sw / np.linalg.norm(sqrt_sw)
+            intercept_dim = _find_smallest_angle(normalized_sw, U)
+            # cancel the regularization for the intercept
+            w[intercept_dim] = - (alpha ** -1)
         c = np.dot(U, self._diag_dot(w, UT_y)) + (alpha ** -1) * y
         G_diag = self._decomp_diag(w, U) + (alpha ** -1)
         if len(y.shape) != 1:
             # handle case where y is 2-d
             G_diag = G_diag[:, np.newaxis]
         return G_diag, c
 
-    def _errors_svd(self, alpha, y, v, U, UT_y):
-        G_diag, c = self._errors_and_values_svd_helper(alpha, y, v, U, UT_y)
+    def _errors_covariance_sparse(self, alpha, y, sqrt_sw, X_mean, v, U, UT_y):
+        G_diag, c = self._errors_and_values_covariance_sparse(
+            alpha, y, sqrt_sw, X_mean, v, U, UT_y)
         return (c / G_diag) ** 2, c
 
-    def _values_svd(self, alpha, y, v, U, UT_y):
-        G_diag, c = self._errors_and_values_svd_helper(alpha, y, v, U, UT_y)
+    def _values_covariance_sparse(self, alpha, y, sqrt_sw, X_mean, v, U, UT_y):
+        G_diag, c = self._errors_and_values_covariance_sparse(
+            alpha, y, sqrt_sw, X_mean, v, U, UT_y)
+        return y - (c / G_diag), c
+
+    def _errors_covariance_dense(self, alpha, y, sqrt_sw, X_mean, v, U, UT_y):
+        G_diag, c = self._errors_and_values_covariance_dense(
+            alpha, y, sqrt_sw, v, U, UT_y)
+        return (c / G_diag) ** 2, c
+
+    def _values_covariance_dense(self, alpha, y, sqrt_sw, X_mean, v, U, UT_y):
+        G_diag, c = self._errors_and_values_covariance_dense(
+            alpha, y, sqrt_sw, v, U, UT_y)
         return y - (c / G_diag), c
```

Comment:
This pattern is repeated many times. Wouldn't it be possible only call the `_errors_and_values` function and do the `squared_errors = (c / G_diag) ** 2` or `predictions = y - (c / G_diag)` transformations only once in the `_RidgeGCV.fit` method?

The name `_errors_and_values` is really misleading. It should probably be renamed to something more either more explicit more generic but less confusing such as `_solve`:

```python
G_diag, dual_coef = _solve(float(alpha), y, sqrt_sw, X_mean, *decomposition)

if error:
   squared_errors = (dual_coef / G_diag) ** 2
   ...
else:
   predictions = y - (dual_coef / G_diag)
   ...
```

Reply:
It would be great to find a better name for `G_diag` but I am not sure which.

Reply:
maybe something like `inv_regularized_gram_diag`? one advandage of `G_diag` is that G is the notation used in the reference

Reply:
Alright. Let's keep `G_diag` then.